### PR TITLE
Expose Filter push down support

### PR DESCRIPTION
**Pull request by arunthirupathi**
For computed/dynamic columns, expression should not be pushed
inside the TableScan. The expression should be applied as a 
FilterNode after the table scan. This change exposes a method to
control what expressions are pushed inside the TableScan.

Test plan - 
Existing tests.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java [hunk 1]
```diff
@@ -265,8 +266,18 @@ public static ConnectorPushdownFilterResult pushdownFilter(
 
         LogicalRowExpressions logicalRowExpressions = new LogicalRowExpressions(rowExpressionService.getDeterminismEvaluator(), functionResolution, functionMetadataManager);
         List<RowExpression> conjuncts = extractConjuncts(decomposedFilter.getRemainingExpression());
-        RowExpression dynamicFilterExpression = extractDynamicConjuncts(conjuncts, logicalRowExpressions);
-        RowExpression remainingExpression = extractStaticConjuncts(conjuncts, logicalRowExpressions);
+        ImmutableList.Builder<RowExpression> dynamicConjuncts = ImmutableList.builder();
+        ImmutableList.Builder<RowExpression> staticConjuncts = ImmutableList.builder();
+        for (RowExpression conjunct : conjuncts) {
+            if (isDynamicFilter(conjunct) || useDynamicFilter(conjunct, tableHandle, columnHandles)) {
```
Comment (pranjalssh) [inline presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:272]: Lets have only one method, which defaults to isDynamicFilter() by default

Comment (pranjalssh) [inline presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:272]: Lets have only one method in the if clause, which is the protected method. It should default to isDynamicFilter().

File: presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java [hunk 2]
```diff
@@ -423,6 +434,17 @@ public PlanNode visitTableScan(TableScanNode tableScan, RewriteContext<Void> con
                     tableScan.getTableConstraints(),
                     pushdownFilterResult.getLayout().getPredicate(),
                     TupleDomain.all());
+
```
Comment (pranjalssh) [inline presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:437]: It'd be good to know when this happens. Discussed offline, we can write this as assert and see if we really hit it

File: presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java [hunk 3]
```diff
@@ -153,6 +151,18 @@ public PlanNode optimize(PlanNode maxSubplan, ConnectorSession session, Variable
         return rewriteWith(new Rewriter(session, idAllocator), maxSubplan);
     }
 
+    // If this method returns true, the filter will not be pushed inside the TableScan node.
+    // This is exposed for dynamic or computed column functionality support.
+    // Consider the following case:
+    // Read a base page from the file. Modify the page to add a new column or update an existing column
```
Comment (pranjalssh) [inline presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:157]: Can we avoid mentioning "page" here, since its a detail that shouldn't come up during planning. You can just mention row

File: presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java [hunk 4]
```diff
@@ -153,6 +151,18 @@ public PlanNode optimize(PlanNode maxSubplan, ConnectorSession session, Variable
         return rewriteWith(new Rewriter(session, idAllocator), maxSubplan);
     }
 
+    // If this method returns true, the filter will not be pushed inside the TableScan node.
```
Comment (pranjalssh) [inline presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:154]: Make this a javadoc. Use params like `expression` instead of "filter"

File: presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java (hunk 1 above)
Comment (arunthirupathi) [reply presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:272]: I believe pushing this inside one method, might have complication.

for isDynamicFilter, we never want them to push down and that should be an invariant. Overriding the method then overrider need to enforce this constraint, that should not be upto the implementation. so I would prefer to keep this as it is.

**Review by highker — APPROVED** [review]

File: presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java (hunk 1 above)
Comment (rschlussel) [reply presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:279]: should we rename the dynamic conjuncts variable if it includes some non-dynamic conjuncts?

Comment (arunthirupathi) [reply presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:279]: I believe they are still dynamic, one is for dynamicFilter and other is dynamic columns. But if there are other names, I can change it.

File: presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java [hunk 5]
```diff
@@ -423,6 +441,12 @@ public PlanNode visitTableScan(TableScanNode tableScan, RewriteContext<Void> con
                     tableScan.getTableConstraints(),
                     pushdownFilterResult.getLayout().getPredicate(),
                     TupleDomain.all());
+
+            RowExpression unenforcedFilter = pushdownFilterResult.getUnenforcedConstraint();
+            if (!TRUE_CONSTANT.equals(unenforcedFilter)) {
```
Comment (rschlussel) [inline presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:446]: I think this should be good, but. I would maybe put it in a separate commit, since it's not related to the rest of the change.  That way if it causes an unexpected issue, we don't have to revert the rest of your changes.

File: presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java (hunk 1 above)
Comment (rschlussel) [reply presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdown.java:279]: ah okay.  I think i just find isDynamicFilter() and useDynamicFilter() to be confusing. useDynamicFilter() sounds like it's not really a dynamic filter, but we're just using that field.  something like `canBeAppliedAtScan()`  might be clearer. not sure if it's too much of a mouthful.
